src/omt/integer_optimizer.cpp

```diff
@@ -0,0 +1,90 @@
+/*********************                                                        */
+/*! \file integer_optimizer.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Yancheng Ou
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ ** in the top-level source directory and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Optimizer for Integer type
+ **/
+
+#include "integer_optimizer.h"
+
+#include "options/smt_options.h"
+#include "smt/smt_engine.h"
+#include "theory/quantifiers/quantifiers_attributes.h"
+#include "theory/smt_engine_subsolver.h"
+
+using namespace CVC4::theory;
+
+namespace CVC4::smt {
+
+std::pair<OptResult, Node> OMTOptimizerInteger::optimize(
+    SmtEngine* parentSMTSolver, Node target, ObjectiveType objType)
+{
+  // linear search for integer goal
+  // the smt engine to which we send intermediate queries
+  // for the linear search.
+  std::unique_ptr<SmtEngine> optChecker =
+      OMTOptimizer::createOptCheckerWithTimeout(parentSMTSolver, false);
+  NodeManager* nm = optChecker->getNodeManager();
+
+  Result intermediateSatResult = optChecker->checkSat();
+  // Model-value of objective (used in optimization loop)
+  Node value;
+  if (intermediateSatResult.isUnknown())
+  {
+    return std::make_pair(OptResult::OPT_UNKNOWN, value);
+  }
+  if (!intermediateSatResult.isSat())
+  {
+    return std::make_pair(OptResult::OPT_UNSAT, value);
+  }
+  // asserts objective > old_value (used in optimization loop)
+  Node increment;
+  Kind incrementalOperator = kind::NULL_EXPR;
+  if (objType == ObjectiveType::OBJECTIVE_MINIMIZE)
+  {
+    // if objective is MIN, then assert optimization_target <
+    // current_model_value
+    incrementalOperator = kind::LT;
+  }
+  else if (objType == ObjectiveType::OBJECTIVE_MAXIMIZE)
+  {
+    // if objective is MAX, then assert optimization_target >
+    // current_model_value
+    incrementalOperator = kind::GT;
+  }
+  // Workhorse of linear search:
+  // This loop will keep incrmenting/decrementing the objective until unsat
+  // When unsat is hit,
+  // the optimized value is the model value just before the unsat call
+  while (intermediateSatResult.isSat())
+  {
+    value = optChecker->getValue(target);
+    Assert(!value.isNull());
+    increment = nm->mkNode(incrementalOperator, target, value);
+    optChecker->assertFormula(increment);
+    intermediateSatResult = optChecker->checkSat();
+  }
+  return std::make_pair(OptResult::OPT_OPTIMAL, value);
+}
+
+std::pair<OptResult, Node> OMTOptimizerInteger::minimize(
```

Comment:
Should this just be the default implementation of `OMTOptimizer::minimize`?

Reply:
Hmm... I think the "less than (LT)" operator and the "LE" operator only works for Integer and Real. So I didn't make it universal... and thus I didn't move OMTOptimizerInteger::optimize to OMTOptimizer either~  
Also using this linear search algorithm on reals might not run into completion -- it keeps finding bad values... I had a testcase for this, but it's removed in this PR since it doesn't run to completion. 

Notice that the testcase for the old framework... not the new refactored one...
`
void testOmtOpenReal()
  {
    Term ub1 = d_slv->mkReal(100);
    Term lb1 = d_slv->mkReal(0);
    Term lb2 = d_slv->mkReal(110);
    Term ub2 = d_slv->mkReal(120);

    Term cost1 = d_slv->mkConst(d_slv->getRealSort(), "cost3"); 
    Term cost2 = d_slv->mkConst(d_slv->getRealSort(), "cost4");
    // 0 <= cost1 <= 100
    d_slv->assertFormula(d_slv->mkTerm(CVC4::api::Kind::LEQ, lb1, cost1));
    d_slv->assertFormula(d_slv->mkTerm(CVC4::api::Kind::LEQ, cost1, ub1));
    // 110 <= cost2 <= 120
    d_slv->assertFormula(d_slv->mkTerm(CVC4::api::Kind::LEQ, lb2, cost2));
    d_slv->assertFormula(d_slv->mkTerm(CVC4::api::Kind::LEQ, cost2, ub2));

    Term cost3 = d_slv->mkTerm(CVC4::api::PLUS, cost1, cost2);

    const ObjectiveType obj_type = OBJECTIVE_MAXIMIZE;
    d_optslv->activateObj(*cost3.d_node, obj_type);

    OptResult r = d_optslv->checkOpt();

    // TS_ASSERT_EQUALS(ub1, ub2);

    TS_ASSERT_EQUALS(d_optslv->objectiveGetValue(), d_nm->mkConst(Rational(220)));
    std::cout << "Result is :" << r << std::endl;
    std::cout << "Optimized value is: " << d_optslv->objectiveGetValue() << std::endl;
  }
`

Reply:
So the body of this method is a one line method:
```
return this->optimize(
      parentSMTSolver, target, ObjectiveType::OBJECTIVE_MINIMIZE);
```
This itself doesn't contain any reference to anything integer specific.  Indeed, it should be the case that target may be integer-specific (?), but I don't see why this method in its current form needs to be here.

Reply:
Yes, this linear search could be used on Reals, just doesn't run to completion sometimes (for bounded cases)... So I would try not to use it on Reals. 
This is the linear search currently in master and it's only for integers. 
I would figure out an approach for LIRA later on...

Reply:
I see, ok, I guess there isn't a need to create OMTOptimizer::optimize for now.

Reply:
Thanks for the review Dr. Andrew!